docs/specs/2fa-grace-period.md

```diff
@@ -0,0 +1,77 @@
+# 2FA Grace Period
+
+- [Abstract](#abstract)
+- [Configuration](#configuration)
+  - [Global Grace Period](#global-grace-period)
+  - [Per-user Grace Period](#per-user-grace-period)
+  - [Customized Authentication Flow](#customized-authentication-flow)
+
+## Abstract
+
+In order to enforce 2FA, we will provide a grace period for users to enroll in 2FA after they have signed up.
+
+After the grace period, the user will be required to enroll in 2FA before they can sign in.
+
+## Configuration
+
+By default, no grace period is provided. User without 2FA must contact admin to login after [2FA mode](./user-model.md#secondary-authenticator) is set to `required`.
+
+For new users, the grace period starts from the time the user is created, whereas for existing users, the grace period starts from when the grace period is enabled.
+
+### Global Grace Period
+
+Global grace period can be enabled for forcing users to enroll in 2FA upon login.
+
+```yaml
+authentication:
+  secondary_authentication_mode: "required"
+  secondary_authentication_rollout:
+    global_grace_period_enabled_since: "2021-01-01T00:00:00Z"
+    global_grace_period_days: 30
+```
+
+### Per-user Grace Period
+
+Regardless of the global grace period, a per-user grace period can be granted for 10 days by default, or a custom duration can be specified:
+
+```yaml
+authentication:
+  secondary_authentication_mode: "required"
+  secondary_authentication_rollout:
+    per_user_grace_period_days: 10
+```
```

Comment:
Remove this section. No need to add this now. Conceptually, it is an argument when calling the new mutation to extend the grace period of a user. Of course, we have validation on the input so that it is not in the past, and it is not too far in the future. I think 90 days is enough.

Can you instead specify the proposed changes to the Admin API here? They are public API that we need to review together.

Reply:
Updated to graphql schema